src/GraphBLAS-sharp/Implementations.fs

```diff
@@ -112,31 +112,1587 @@ type CSRMatrix<'a when 'a : struct and 'a : equality>(csrTuples: CSRFormat<'a>)
     override this.Reduce a = failwith "Not Implemented"
     override this.T = failwith "Not Implemented"
 
+and COOMatrix<'a when 'a : struct and 'a : equality>(elements: (int * int * 'a)[], rowCount: int, columnCount: int) =
+    inherit Matrix<'a>(rowCount, columnCount)
+
+    let mutable rows, columns, values = elements |> Array.unzip3
+    member this.Rows with get() = rows
+    member this.Columns with get() = columns
+    member this.Values with get() = values
+    member this.Elements with get() = elements
+
+    override this.Mask = Some <| Mask2D(Array.zip this.Rows this.Columns, this.RowCount, this.ColumnCount, false)
+    override this.Complemented = Some <| Mask2D(Array.zip this.Rows this.Columns, this.RowCount, this.ColumnCount, true)
+
+    member private this.SetCOOMatrix
+        (mask: Mask2D)
+        (matrix: COOMatrix<'a>) =
+        let newElements = matrix.Elements |> Array.filter (fun (i, j, _) -> mask.[i, j])
+        let elementsToSet, elementsToAppend =
+            newElements
+            |> Array.partition (
+                fun (i, j, _) ->
+                this.Elements
+                |> Array.exists (fun (u, v, _) -> (u, v) = (i, j)))
+
+        let workflow =
+            opencl {
+                let command =
```

Comment:
Do not declare all these variables inside computation expression. This is a cause of deep code nesting. You could write only the last two lines inside it.